<br>
<font face="黑体"> Version of 20 interval of y  (unit: 5000dollars/ num of data:20000)

In [485]:
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd 

from keras.models import Sequential
from keras.layers import Dense, Activation
from keras.optimizers import SGD

from keras.datasets import mnist

from keras.utils import np_utils

df = pd.read_excel('raw_used_car_data.xlsx')

In [486]:
df.head()

ID     make    model  months_old  power sale_type  num_owners gear_type  \
0  97860  Porsche      911       240.0  210.0   classic         3.0    manual   
1  27821     Ford  Mustang        54.0  487.0      used         1.0    manual   
2  97801  Porsche      911       358.0  220.0      used         NaN    manual   
3  98251  Porsche      911        14.0  368.0      used         1.0    manual   
4  98250  Porsche      911         3.0  515.0      used         NaN       NaN   

  fuel_type       kms     price  
0  gasoline  202000.0  999999.0  
1  gasoline   30000.0  685000.0  
2  gasoline   56300.0  555555.0  
3  gasoline    2800.0  470000.0  
4  gasoline      10.0  450000.0

In [487]:
##nurmeric
df_t = pd.read_excel('training_x.xlsx')

In [488]:
df_t.head()

ID  make  months_old  power  sale_type  num_owners  gear_type  \
0  39508    24          13     92          5           1          2   
1  71262    10         121     51          5           2          2   
2  11149     4          14     85          5           1          2   
3  56224    35          52    103          5           1          1   
4  73615    10         121     44          5           1          2   

   fuel_type     kms  price  
0          5   11104  14295  
1          5   90000   5500  
2          2   18719  22900  
3          2  109000  14990  
4          5   66371   3900

In [489]:
##x data
x_train= np.array([df_t['make'] , df_t['months_old'] , df_t['power'] , df_t['sale_type'] , df_t['num_owners'] ,  df_t['gear_type'] , 
            df_t['fuel_type'] , df_t['kms']]) 
x_train=x_train.T
print (x_train)

[[    24     13     92 ...      2      5  11104]
 [    10    121     51 ...      2      5  90000]
 [     4     14     85 ...      2      2  18719]
 ...
 [     8     79     81 ...      2      2 198000]
 [    27      8     96 ...      2      2      1]
 [    24      4    100 ...      2      2   1510]]


In [490]:
len(x_train)

19998

In [491]:
df_t.corr()

ID      make  months_old     power  sale_type  num_owners  \
ID          1.000000  0.670539    0.088192 -0.159060   0.004873    0.012924   
make        0.670539  1.000000    0.091071 -0.212742   0.032343    0.011883   
months_old  0.088192  0.091071    1.000000  0.039280   0.161721    0.258894   
power      -0.159060 -0.212742    0.039280  1.000000   0.012678    0.087850   
sale_type   0.004873  0.032343    0.161721  0.012678   1.000000    0.029887   
num_owners  0.012924  0.011883    0.258894  0.087850   0.029887    1.000000   
gear_type   0.216064  0.228619    0.070792 -0.524019   0.016469   -0.031207   
fuel_type   0.181527  0.123525    0.102184  0.041955   0.009106    0.054908   
kms         0.045880  0.077667    0.756328  0.038164   0.160117    0.187247   
price      -0.126523 -0.182194   -0.292732  0.659406  -0.087729   -0.002955   

            gear_type  fuel_type       kms     price  
ID           0.216064   0.181527  0.045880 -0.126523  
make         0.228619   0.123525  0.077667 -0.182194  
months_old   0.070792   0.102184  0.756328 -0.292732  
power       -0.524019   0.041955  0.038164  0.659406  
sale_type    0.016469   0.009106  0.160117 -0.087729  
num_owners  -0.031207   0.054908  0.187247 -0.002955  
gear_type    1.000000   0.063857  0.040585 -0.410623  
fuel_type    0.063857   1.000000 -0.104611 -0.019004  
kms          0.040585  -0.104611  1.000000 -0.302651  
price       -0.410623  -0.019004 -0.302651  1.000000

In [492]:
## put y in [1*20]
df_tt = pd.read_excel('training_y.xlsx')

In [493]:
df_tt.head()

1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  19  \
0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
1   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
2   0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
3   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
4   1   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   

   20  
0   0  
1   0  
2   0  
3   0  
4   0

In [494]:
##y data
y_train= np.array([df_tt[1] , df_tt[2] , df_tt[3] , df_tt[4] , df_tt[5] , df_tt[6] , df_tt[7] , df_tt[8] , df_tt[9]  , df_tt[10] 
             , df_tt[11] , df_tt[12] , df_tt[13] , df_tt[14] , df_tt[15] , df_tt[16] , df_tt[17] , df_tt[18]  , df_tt[19] , df_tt[20]])
y_train=y_train.T
print (y_train)

[[0 0 1 ... 0 0 0]
 [0 1 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 1 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [495]:
len(y_train)

19998

In [496]:
## actual y
y_actual_train= np.array([df_t['price'] ]) 
y_actual_train=y_actual_train.T
print (y_actual_train)

[[14295]
 [ 5500]
 [22900]
 ...
 [ 8995]
 [22610]
 [16100]]


In [497]:
##test data
X = x_train[87]

In [498]:
X

array([    4,    41,   190,     5,     1,     1,     2, 39006])

In [499]:
Y = y_actual_train[87]

In [500]:
Y

array([49900])

Organize the input/output 

In [501]:
x_train = x_train.reshape(19998, 8)

In [502]:
y_train = y_train.reshape(19998, 20)

In [503]:
##normalize
x_train = x_train-x_train.min()
x_train = x_train/x_train.max()
x_train.min(), x_train.max()

(0.0, 1.0)

In [504]:
from keras.models import Sequential
from keras.layers import Dense, Activation
from keras.optimizers import SGD
from keras.models import Model
from keras.layers import Input
from keras.layers import concatenate, add
%env KERAS_BACKEND=tensorflow

env: KERAS_BACKEND=tensorflow


In [505]:
f_1 = Dense(10, activation='sigmoid')
f_2 = Dense(10, activation='relu')
f_3 = Dense(10, activation='sigmoid')
f_4 = Dense(10, activation='tanh')
f_5 = Dense(10, activation='relu')
f_6 = Dense(10, activation='sigmoid')
f_7 = Dense(20, activation='softmax')

x = Input(shape=(8,))

h_1 = f_1(x)
h_2 = f_2(x)
c1 = concatenate([h_1, h_2])
h_3 = f_3(c1)
h_4 = f_4(h_3)
h_5 = f_5(h_3)
h_6 = f_6(h_3)
c2 = concatenate([h_4, h_5, h_6])
y = f_7(c2)

In [506]:
model = Model(x, y)

In [507]:
model.summary()

__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_22 (InputLayer)           (None, 8)            0                                            
__________________________________________________________________________________________________
dense_220 (Dense)               (None, 10)           90          input_22[0][0]                   
__________________________________________________________________________________________________
dense_221 (Dense)               (None, 10)           90          input_22[0][0]                   
__________________________________________________________________________________________________
concatenate_27 (Concatenate)    (None, 20)           0           dense_220[0][0]                  
                                                                 dense_221[0][0]                  
__________

In [508]:
model.compile(loss='categorical_crossentropy', optimizer=SGD(lr=0.08), metrics=['accuracy'])
model.fit(x_train, y_train, batch_size=50, epochs=50, verbose=1, validation_split=0.2)

Train on 15998 samples, validate on 4000 samples
Epoch 1/50
15998/15998 [==============================] - 3s 195us/step - loss: 2.1674 - acc: 0.2283 - val_loss: 2.1295 - val_acc: 0.2450
Epoch 2/50
15998/15998 [==============================] - 1s 52us/step - loss: 2.1189 - acc: 0.2259 - val_loss: 2.1323 - val_acc: 0.2450
Epoch 3/50
15998/15998 [==============================] - 1s 63us/step - loss: 2.1139 - acc: 0.2254 - val_loss: 2.1233 - val_acc: 0.2450
Epoch 4/50
15998/15998 [==============================] - 1s 52us/step - loss: 2.1035 - acc: 0.2372 - val_loss: 2.1156 - val_acc: 0.2393
Epoch 5/50
15998/15998 [==============================] - 1s 50us/step - loss: 2.0767 - acc: 0.2543 - val_loss: 2.0715 - val_acc: 0.2490
Epoch 6/50
15998/15998 [==============================] - 1s 57us/step - loss: 2.0214 - acc: 0.2635 - val_loss: 2.0251 - val_acc: 0.2632
Epoch 7/50
15998/15998 [==============================] - 1s 53us/step - loss: 1.9846 - acc: 0.2662 - val_loss: 2.0084 - val_acc

<br>
<font face="黑体"> Conclusion
    
1/ the accuracy of training machine~27%<br>
2/the accuracy of test~27%<br>